### PR TITLE
Add pattern to canonicalize for loop bounds

### PR DESCRIPTION
**Pull request by bondhugula**
- add pattern to canonicalize affine.for loop bounds (using
  canonicalizeMapAndOperands)
- rename AffineForLoopBoundFolder -> AffineForOpBoundFolder for consistency

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

File: lib/Dialect/AffineOps/AffineOps.cpp [hunk 1]
```diff
@@ -1339,11 +1339,53 @@ struct AffineForLoopBoundFolder : public OpRewritePattern<AffineForOp> {
...
+    return matchSuccess();
+  }
+};
+
 } // end anonymous namespace
 
 void AffineForOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
                                               MLIRContext *context) {
-  results.insert<AffineForLoopBoundFolder>(context);
+  results.insert<AffineForOpBoundFolder>(context);
```
Comment (River707) [inline lib/Dialect/AffineOps/AffineOps.cpp]: 'insert' is variadic <..., ...>

Comment (bondhugula) [reply lib/Dialect/AffineOps/AffineOps.cpp]: Thanks, changed.

File: lib/Dialect/AffineOps/AffineOps.cpp [hunk 2]
```diff
@@ -1354,11 +1354,53 @@ struct AffineForLoopBoundFolder : public OpRewritePattern<AffineForOp> {
...
   }
 };
+
+// This is a pattern to canonicalize affine for op loop bounds.
+struct AffineForOpBoundCanonicalizer : public OpRewritePattern<AffineForOp> {
+  using OpRewritePattern<AffineForOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(AffineForOp forOp,
+                                     PatternRewriter &rewriter) const override {
+    SmallVector<Value *, 4> lbOperands(forOp.getLowerBoundOperands());
```
Comment (nicolasvasilache) [inline lib/Dialect/AffineOps/AffineOps.cpp]: How about this?

```
SmallVector<Value *, 4> lbOperands(forOp.getLowerBoundOperands());
SmallVector<Value *, 4> ubOperands(forOp.getUpperBoundOperands());

auto lbMap = forOp.getLowerBoundMap();
auto ubMap = forOp.getUpperBoundMap();
auto tmpLbMap = lbMap;
auto tmpUbMap = ubMap;

canonicalizeMapAndOperands(&lbMap, &lbOperands);
canonicalizeMapAndOperands(&ubMap, &ubOperands);

// If none of lbMap and ubMap are update by canonicalization, 
// don't rewrite.
if (lbMap == tmpLbMap && ubMap == tmpUbMap)
  return matchFailure();
      
forOp.setLowerBound(lbOperands, lbMap);
forOp.setUpperBound(ubOperands, ubMap);
rewriter.updatedRootInPlace(forOp);
return matchSuccess();
```

File: test/AffineOps/canonicalize.mlir [hunk 3]
```diff
@@ -424,3 +424,33 @@ func @fold_empty_loop() {
   }
   return
 }
+
+// -----
+
+// CHECK-DAG: [[LBMAP:#map[0-9]+]] = ()[s0] -> (0, s0)
+// CHECK-DAG: [[UBMAP:#map[0-9]+]] = ()[s0] -> (1024, s0 + s0)
```
Comment (nicolasvasilache) [inline test/AffineOps/canonicalize.mlir]: shouldn't this fold to `2 * s0` ?

File: lib/Dialect/AffineOps/AffineOps.cpp (hunk 2 above)
Comment (bondhugula) [reply lib/Dialect/AffineOps/AffineOps.cpp]: Yes, this works too, and is simpler - but we don't need to update both lb and ub operands if only one of them had been changed. Updated to it with a slight change. Thanks.

File: test/AffineOps/canonicalize.mlir (hunk 3 above)
Comment (bondhugula) [reply test/AffineOps/canonicalize.mlir]: We didn't include such a simplification (or even say s0 - s0) in AffineExpr::simplifyAdd; so this won't fold. I think AffineExpr::simplifyAdd should be updated to at least fold diff of identical expressions by construction.

File: lib/Dialect/AffineOps/AffineOps.cpp [hunk 4]
```diff
@@ -1354,11 +1354,51 @@ struct AffineForLoopBoundFolder : public OpRewritePattern<AffineForOp> {
...
+    auto lbMap = forOp.getLowerBoundMap();
+    auto ubMap = forOp.getUpperBoundMap();
+    auto prevLbMap = lbMap;
+    auto prevUbMap = ubMap;
+
+    canonicalizeMapAndOperands(&lbMap, &lbOperands);
+    canonicalizeMapAndOperands(&ubMap, &ubOperands);
+
+    // Any canonicalization change always leads to updated map(s).
+    bool canonicalized = false;
```
Comment (nicolasvasilache) [inline lib/Dialect/AffineOps/AffineOps.cpp]: Early exit if none has changed still looks preferable to me:
```
if (lbMap == prevLbMap && ubMap == prevUbMap)
  return matchFailure;
if (lbMap != prevLbMap)
  forOp.setLowerBound(lbOperands, lbMap);
if (ubMap != prevUbMap)
  forOp.setUpperBound(ubOperands, ubMap);
rewriter.updatedRootInPlace(forOp);
return matchSuccess();
```

Comment (bondhugula) [reply lib/Dialect/AffineOps/AffineOps.cpp]: Done.

File: lib/Dialect/AffineOps/AffineOps.cpp [hunk 5]
```diff
@@ -1354,11 +1354,46 @@ struct AffineForLoopBoundFolder : public OpRewritePattern<AffineForOp> {
...
+    auto lbMap = forOp.getLowerBoundMap();
+    auto ubMap = forOp.getUpperBoundMap();
+    auto prevLbMap = lbMap;
+    auto prevUbMap = ubMap;
+
+    canonicalizeMapAndOperands(&lbMap, &lbOperands);
+    canonicalizeMapAndOperands(&ubMap, &ubOperands);
+
+    // Any canonicalization change always leads to updated map(s).
+    if (lbMap == prevLbMap && ubMap == prevUbMap) {
```
Comment (River707) [inline lib/Dialect/AffineOps/AffineOps.cpp]: Let's drop the trivial braces on these ifs.

Comment (bondhugula) [reply lib/Dialect/AffineOps/AffineOps.cpp]: Done.

**Comment by bondhugula** [comment]
Fixed merge conflicts. @nicolasvasilache @River707

**Comment by joker-eph** [comment]
I see failures on the examples tests, for example:

```
mlir/examples/Linalg/Linalg3/Example.cpp:65:12: error: CHECK: expected string not found in input
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[N]]) {
           ^
<stdin>:10:2: note: scanning from here
 affine.for %arg3 = 0 to %1 {
 ^
<stdin>:10:2: note: with "N" equal to "1"
 affine.for %arg3 = 0 to %1 {
 ^
```

**Comment by bondhugula** [comment]
> I see failures on the examples tests, for example:
> 
> ```
> mlir/examples/Linalg/Linalg3/Example.cpp:65:12: error: CHECK: expected string not found in input
>  // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[N]]) {
>            ^
> <stdin>:10:2: note: scanning from here
>  affine.for %arg3 = 0 to %1 {
>  ^
> <stdin>:10:2: note: with "N" equal to "1"
>  affine.for %arg3 = 0 to %1 {
>  ^
> ```

How do I reproduce these? I already have the default cmake conf (per README.md) with LLVM_BUILD_EXAMPLES=ON, and check-mlir passes for me.

**Comment by joker-eph** [comment]
> How do I reproduce these? I already have the default cmake conf (per README.md) with LLVM_BUILD_EXAMPLES=ON, and check-mlir passes for me.

That's why I kicked the "Kokoro" builds earlier to see if this is an internal only issue, and these builds seem to be passing. I'll involve @nicolasvasilache to debug this! 
Sorry for the delay this is causing!

**Comment by bondhugula** [comment]
> > How do I reproduce these? I already have the default cmake conf (per README.md) with LLVM_BUILD_EXAMPLES=ON, and check-mlir passes for me.
> 
> That's why I kicked the "Kokoro" builds earlier to see if this is an internal only issue, and these builds seem to be passing. I'll involve @nicolasvasilache to debug this!
> Sorry for the delay this is causing!

This one is pretty easy to fix. As a result of canonicalization, (d0) -> (d0) here turns into ()[s0] -> (s0), and (s0) -> (s0)(%[[N]]) will just be printed as %[[N]]. 

But I wanted a way to have these missing tests run - so that I could catch these and any more on my own.

**Comment by nicolasvasilache** [comment]
Took a quick look, `mlir/examples/Linalg/Linalg3/Example.cpp` and `mlir/examples/Linalg/Linalg4/Example.cpp` should be very easy to fix "with a blindfold". 
They all seem to be instances of map + parameter ub that should just become the parameter:
```
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[N]]) {
           ^
<stdin>:93:2: note: scanning from here
 affine.for %arg3 = 0 to %1 {
 ^
<stdin>:93:2: note: with "N" equal to "1"
 affine.for %arg3 = 0 to %1 {
```
and
```
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[M]]) step 8 {
           ^
<stdin>:32:2: note: scanning from here
 affine.for %arg3 = 0 to %0 step 8 {
 ^
<stdin>:32:2: note: with "M" equal to "0"
 affine.for %arg3 = 0 to %0 step 8 {
 ^
```

Maybe take a quick crack at them in this CL?
Otherwise I can pick it up Monday when I know how we proceed with making changes to a CL from another user.

**Comment by bondhugula** [comment]
Fixing it isn't the issue! I was asking how I could run these tests in the first place? (Please see above - they aren't running as part of chrck-mlir.)  would like to avoid a situation where some more test cases to update are found after a few days.

**Comment by jpienaar** [comment]
> Fixing it isn't the issue! I was asking how I could run these tests in the first place? (Please see above - they aren't running as part of chrck-mlir.) would like to avoid a situation where some more test cases to update are found after a few days.

It seems there is a pattern missing that groups these into a convenient cmake target, but you can run them individually. E.,g, 

ninja linalg-example-3
./bin/linalg-example-3 | ./bin/FileCheck ../mlir/examples/Linalg/Linalg3/Example.cpp

**Comment by bondhugula** [comment]
> > Fixing it isn't the issue! I was asking how I could run these tests in the first place? (Please see above - they aren't running as part of chrck-mlir.) would like to avoid a situation where some more test cases to update are found after a few days.
> 
> It seems there is a pattern missing that groups these into a convenient cmake target, but you can run them individually. E.,g,
> 
> ninja linalg-example-3
> ./bin/linalg-example-3 | ./bin/FileCheck ../mlir/examples/Linalg/Linalg3/Example.cpp

Thanks very much! This is exactly what I was looking for! Fixed now.

**Comment by joker-eph** [comment]
We still see the other Linalg example failing apparently:

```

/build/work/a961fabe6974af72c61c07bcbfdfb8f5a2ba/google3/runfiles/google3/third_party/llvm/llvm/projects/google_mlir/examples/Linalg/Linalg4/Example.cpp:71:12: error: CHECK: expected string not found in input
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[M]]) step 8 {
           ^
<stdin>:7:2: note: scanning from here
 affine.for %arg3 = 0 to %0 step 8 {
 ^
<stdin>:7:2: note: with "M" equal to "0"
 affine.for %arg3 = 0 to %0 step 8 {
 ^
/build/work/a961fabe6974af72c61c07bcbfdfb8f5a2ba/google3/runfiles/google3/third_party/llvm/llvm/projects/google_mlir/examples/Linalg/Linalg4/Example.cpp:101:12: error: CHECK: expected string not found in input
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[M]]) step 8 {
           ^
<stdin>:32:2: note: scanning from here
 affine.for %arg3 = 0 to %0 step 8 {
 ^
<stdin>:32:2: note: with "M" equal to "0"
 affine.for %arg3 = 0 to %0 step 8 {
 ^
/build/work/a961fabe6974af72c61c07bcbfdfb8f5a2ba/google3/runfiles/google3/third_party/llvm/llvm/projects/google_mlir/examples/Linalg/Linalg4/Example.cpp:135:12: error: CHECK: expected string not found in input
 // CHECK: affine.for %{{.*}} = 0 to (d0) -> (d0)(%[[M]]) step 8 {
           ^
<stdin>:54:2: note: scanning from here
 affine.for %arg3 = 0 to %0 step 8 {
 ^
<stdin>:54:2: note: with "M" equal to "0"
 affine.for %arg3 = 0 to %0 step 8 {
 ^
```

**Comment by bondhugula** [comment]
> We still see the other Linalg example failing apparently:
> 

@joker-eph Fixed now.

**Pull request closed without merging**